polar-core/src/data_filtering.rs

```diff
@@ -0,0 +1,799 @@
+                for constraint in &fetch_request.constraints {
+                    let op = match constraint.kind {
+                        ConstraintKind::Eq => "=".to_owned(),
+                        ConstraintKind::In => "in".to_owned(),
+                        ConstraintKind::Contains => "contains".to_owned(),
+                    };
+                    let field = constraint.field.clone();
+                    let value = match &constraint.value {
+                        ConstraintValue::Term(t) => t.to_polar(),
+                        ConstraintValue::Field(f) => format!("FIELD({})", f),
+                        ConstraintValue::Ref(r) => {
+                            let mut s = "REF(".to_owned();
+                            if let Some(field) = &r.field {
+                                s.push_str(&format!("field {} of ", field));
+                            }
+                            s.push_str(&format!("result {})", r.result_id));
+                            s
+                        }
+                    };
+                    eprintln!("          {} {} {}", field, op, value);
+                }
+            }
+        }
+    }
+}
+
+pub type Types = HashMap<String, HashMap<String, Type>>;
+pub type PartialResults = Vec<ResultEvent>;
+
+#[derive(Debug)]
+struct VarInfo {
+    cycles: Vec<(Symbol, Symbol)>,                      // x = y
+    types: Vec<(Symbol, String)>,                       // x matches XClass
+    eq_values: Vec<(Symbol, Term)>,                     // x = 1;
+    contained_values: Vec<(Term, Symbol)>,              // 1 in x
+    field_relationships: Vec<(Symbol, String, Symbol)>, // x.a = y
+    in_relationships: Vec<(Symbol, Symbol)>,            // x in y
+}
+
+// @TODO(steve): Better way to handle these checks than just unwraps and asserts.
+
+fn process_result(exp: &Operation) -> VarInfo {
+    let mut var_info = VarInfo {
+        cycles: vec![],
+        types: vec![],
+        eq_values: vec![],
+        contained_values: vec![],
+        field_relationships: vec![],
+        in_relationships: vec![],
+    };
+    process_exp(&mut var_info, exp);
+    var_info
+}
+
+fn dot_var(var_info: &mut VarInfo, var: Term, field: &Term) -> Symbol {
+    let mut var = var;
+    while let Ok(Operation {
+        operator: Operator::Dot,
+        args,
+    }) = var.value().as_expression()
+    {
+        var = Term::new_temporary(Value::Variable(dot_var(
+            var_info,
+            args[0].clone(),
+            &args[1],
+        )))
+    }
+    // TODO(steve): There's a potential name clash here which would be bad. Works for now.
+    // but should probably generate this var better.
+    let sym = var.value().as_symbol().unwrap();
+    let field_str = field.value().as_string().unwrap();
+    let new_var = Symbol::new(&format!("{}_dot_{}", sym.0, field_str));
+
+    // Record the relationship between the vars.
+    var_info
```

Comment:
It seems like this function just produces the variable name but it actually also updates `var_info`. Any way to keep this in `process_exp`?